src/yz_solrq.erl

```diff
@@ -0,0 +1,817 @@
+
+        end,
+    ?PULSE_DEBUG("drain.  NewState: ~p~n", [debug_state(NewState)]),
+    {noreply, NewState};
+
+
+%% @doc The fuse for the specified index has blown.
+handle_cast({blown_fuse, Index}, State) ->
+    {noreply, handle_blown_fuse(Index, State)};
+
+%%
+%% @doc Clear the fuse_blown state on the IndexQ associated with the supplied Index
+%%      Resume any batches that may need to proceed.
+%% @end
+%%
+handle_cast({healed_fuse, Index}, State) ->
+    IndexQ = get_indexq(Index, State),
+    NewIndexQ = maybe_request_worker(Index, maybe_start_timer(Index, IndexQ#indexq{fuse_blown = false})),
+    {noreply, update_indexq(Index, NewIndexQ, State)};
+
+%%
+%% @doc     Handle the batch_complete message.
+%%
+%%          The batch_complete message is sent via the batch_complete/3 function
+%%          in this module, which is called by a solrq_helper when a batch has
+%%          been delivered to Solr (or has failed to have been delivered).
+%%          If the batch was successfully sent to Solr, the Result field of the
+%%          batch_complete message matches {ok, NumDelivered}.  NumDelivered is
+%%          the number of messages that were successfully delivered to Solr.
+%%          Otherwise, it matches the tuple {retry, NumDelievered, Undelivered},
+%%          where NumDelivered is as above, and Undelivered is a list of messages
+%%          that have not been delivered (a subset of what was sent for delivery)
+%%          and should be retried.  This handler will decrement the all_queues_len
+%%          field on the solrq state record by the supplied NumDelievered value
+%%          thus potentially unblocking any vnodes waiting on this solrq instance,
+%%          if the number of queued messages are above the high water mark.
+%% @end
+%%
+handle_cast({batch_complete, Index, {NumDelivered, Result}}, #state{all_queue_len = AQL} = State) ->
+    ?PULSE_DEBUG("batch_complete.  State: ~p~n", [debug_state(State)]),
+    IndexQ = get_indexq(Index, State),
+    State1 = handle_batch(Index, IndexQ#indexq{pending_helper = false, in_flight_len = 0}, Result, State),
+    NewState = maybe_unblock_vnodes(State1#state{all_queue_len = AQL - NumDelivered}),
+    ?PULSE_DEBUG("batch_complete.  NewState: ~p~n", [debug_state(NewState)]),
+    {noreply, NewState};
+
+%%
+%% @doc     Handle the drain_complete message.
+%%
+%%          The drain_complete message is sent from the yz_solrq_drain_fsm, once
+%%          all the solrqs have completed their drains.  This is a signal to resume
+%%          batching, by moving the aux queue to the main queue, setting the drain
+%%          flag to false (finally), and kick-starting the indexq.
+%% @end
+%%
+handle_cast(drain_complete, #state{indexqs = IndexQs} = State) ->
+    NewIndexQs = dict:fold(
+        fun(Index, #indexq{queue = Queue, queue_len = QueueLen, aux_queue = AuxQueue} = IndexQ, IndexQAccum) ->
+            NewIndexQ = IndexQ#indexq{
+                queue = queue:join(Queue, AuxQueue),
```

Comment:
Shouldn't AuxQueue go before Queue?  In general, Queue should be empty when we receive a drain_complete, and there should be an invariant to that effect, but I am not sure how easy it is to guarantee that.

If not, it seems possible to re-arrange the order of Solr updates during a drain, which could cause strange behavior (e.g., a delete before a write, or conversely), though at this layer, any anomalies might be papered over by AAE.